fencing/admin.c

```diff
@@ -337,32 +364,23 @@ show_history(stonith_t *st, const char *target, int timeout, int quiet,
 
         if (quiet || !verbose) {
             continue;
-        } else if (hp->action == NULL) {
-            action_s = strdup("unknown");
-        } else if (hp->action[0] != 'r') {
-            action_s = crm_concat("turn", hp->action, ' ');
-        } else {
-            action_s = strdup(hp->action);
         }
 
         if (hp->state == st_failed) {
+            action_s = fence_action_str(hp->action);
             printf("%s failed to %s node %s on behalf of %s from %s at %s\n",
                    hp->delegate ? hp->delegate : "We", action_s, hp->target,
                    hp->client, hp->origin, ctime(&complete));
 
-        } else if (hp->state == st_done && hp->delegate) {
-            printf("%s was able to %s node %s on behalf of %s from %s at %s\n",
-                   hp->delegate, action_s, hp->target,
-                   hp->client, hp->origin, ctime(&complete));
-
         } else if (hp->state == st_done) {
-            printf("We were able to %s node %s on behalf of %s from %s at %s\n",
-                   action_s, hp->target, hp->client, hp->origin, ctime(&complete));
+            print_fence_event(latest);
+
         } else {
             /* ocf:pacemaker:controld depends on "wishes to" being
              * in this output, when used with older versions of DLM
              * that don't report stateful_merge_wait
              */
+            action_s = fence_action_str(hp->action);
```

Comment:
didn't see the free for that

Reply:
the free() is after the if-else block

Reply:
sry - misled by the block without action_s - but as free(NULL) is defined as NOP ...